server/channels/store/sqlstore/post_store.go

```diff
@@ -841,19 +841,25 @@ func (s *SqlPostStore) GetSingle(id string, inclDeleted bool) (*model.Post, erro
 }
 
 type etagPosts struct {
-	Id       string
-	UpdateAt int64
+	Id                 string
+	UpdateAt           int64
+	MembershipUpdateAt int64
 }
 
 //nolint:unparam
 func (s *SqlPostStore) InvalidateLastPostTimeCache(channelId string) {
 }
 
 //nolint:unparam
-func (s *SqlPostStore) GetEtag(channelId string, allowFromCache, collapsedThreads bool) string {
-	q := s.getQueryBuilder().Select("Id", "UpdateAt").From("Posts").Where(sq.Eq{"ChannelId": channelId}).OrderBy("UpdateAt DESC").Limit(1)
+func (s *SqlPostStore) GetEtag(channelId, userId string, allowFromCache, collapsedThreads bool) string {
+	q := s.getQueryBuilder().Select("Id", "UpdateAt").From("Posts").Where(sq.Eq{"ChannelId": channelId}).Limit(1)
 	if collapsedThreads {
-		q.Where(sq.Eq{"RootId": ""})
+		q = q.LeftJoin("ThreadMemberships ON ThreadMemberships.PostId = Posts.Id AND ThreadMemberships.UserId=?", userId).
+			Column("COALESCE(ThreadMemberships.LastUpdated, 0) AS MembershipUpdateAt").
+			Where(sq.Eq{"RootId": ""}).
+			OrderBy("GREATEST(COALESCE(ThreadMemberships.LastUpdated, 0), UpdateAt) DESC")
```

Comment:
This could potentially add some performance overhead if there are too many cache invalidations, since this gets executed for every channel load to compute the etag.

Could you please run an `EXPLAIN ANALYZE` of the older vs newer query with CRT on? Just want to see what sort of difference we are getting on the query plan.